src/iop/lut3d.c

```diff
@@ -0,0 +1,1023 @@
+  int kernel_lut3d_trilinear;
+  int kernel_lut3d_pyramid;
+} dt_iop_lut3d_global_data_t;
+
+const char *name()
+{
+  return _("lut 3D");
+}
+
+int flags()
+{
+  return IOP_FLAGS_INCLUDE_IN_STYLES | IOP_FLAGS_SUPPORTS_BLENDING;
+}
+
+int groups()
+{
+  return IOP_GROUP_COLOR;
+}
+
+int default_colorspace(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  return iop_cs_rgb;
+}
+
+// From `HaldCLUT_correct.c' by Eskil Steenberg (http://www.quelsolaar.com) (BSD licensed)
+void correct_pixel_trilinear(float *input, float *output, const float *const clut, const uint8_t level)
+{
+  int red, green, blue, i, j;
+  float tmp[6];
+  const int level2 = level * level;
+
+  for(int c = 0; c < 3; ++c) input[c] = input[c] < 0.0f ? 0.0f : (input[c] > 1.0f ? 1.0f : input[c]);
+  red = input[0] * (float)(level - 1);
+  if(red > level - 2) red = level - 2;
+  if(red < 0) red = 0;
+  green = input[1] * (float)(level - 1);
+  if(green > level - 2) green = level - 2;
+  if(green < 0) green = 0;
+  blue = input[2] * (float)(level - 1);
+  if(blue > level - 2) blue = level - 2;
+  if(blue < 0) blue = 0;
+
+  const float r = input[0] * (float)(level - 1) - red; // delta red
+  const float g = input[1] * (float)(level - 1) - green; // delta green
+  const float b = input[2] * (float)(level - 1) - blue; // delta blue
+
+// indexes of P000 to P111 in clut
+  const int color = red + green * level + blue * level * level;
+  i = color * 3;  // P000
+  j = (color + 1) * 3;  // P100
+
+  tmp[0] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[1] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[2] = clut[i] * (1 - r) + clut[j] * r;
+
+  i = (color + level) * 3;  // P010
+  j = (color + level + 1) * 3;  //P110
+
+  tmp[3] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[4] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[5] = clut[i] * (1 - r) + clut[j] * r;
+
+  output[0] = tmp[0] * (1 - g) + tmp[3] * g;
+  output[1] = tmp[1] * (1 - g) + tmp[4] * g;
+  output[2] = tmp[2] * (1 - g) + tmp[5] * g;
+
+  i = (color + level2) * 3;  // P001
+  j = (color + level2 + 1) * 3;  // P101
+
+  tmp[0] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[1] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[2] = clut[i] * (1 - r) + clut[j] * r;
+
+  i = (color + level + level2) * 3;  // P011
+  j = (color + level + level2 + 1) * 3;  // P111
+
+  tmp[3] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[4] = clut[i++] * (1 - r) + clut[j++] * r;
+  tmp[5] = clut[i] * (1 - r) + clut[j] * r;
+
+  tmp[0] = tmp[0] * (1 - g) + tmp[3] * g;
+  tmp[1] = tmp[1] * (1 - g) + tmp[4] * g;
+  tmp[2] = tmp[2] * (1 - g) + tmp[5] * g;
+
+  output[0] = output[0] * (1 - b) + tmp[0] * b;
+  output[1] = output[1] * (1 - b) + tmp[1] * b;
+  output[2] = output[2] * (1 - b) + tmp[2] * b;
+}
+// from OpenColorIO
+// https://github.com/imageworks/OpenColorIO/blob/master/src/OpenColorIO/ops/Lut3D/Lut3DOp.cpp
+void correct_pixel_tetrahedral(float *input, float *output, const float *const clut, const uint8_t level)

```

Comment:
` void correct_pixel_tetrahedral(float *const input, float *const output, const float *const restrict clut, const uint8_t level) `

`restrict` avoid fake aliasing

Reply:
good to know. done